actors/miner/src/expiration_queue.rs

```diff
@@ -776,46 +777,64 @@ impl<'db, BS: Blockstore> ExpirationQueue<'db, BS> {
         sector_size: SectorSize,
         sectors: &[SectorOnChainInfo],
     ) -> anyhow::Result<Vec<SectorExpirationSet>> {
-        let mut declared_expirations = BTreeMap::<ChainEpoch, bool>::new();
+        if sectors.is_empty() {
+            return Ok(Vec::new());
+        }
+
         let mut sectors_by_number = BTreeMap::<u64, &SectorOnChainInfo>::new();
         let mut all_remaining = BTreeSet::<u64>::new();
 
         for sector in sectors {
-            let q_expiration = self.quant.quantize_up(sector.expiration);
-            declared_expirations.insert(q_expiration, true);
             all_remaining.insert(sector.sector_number);
             sectors_by_number.insert(sector.sector_number, sector);
         }
 
-        let mut expiration_groups =
-            Vec::<SectorExpirationSet>::with_capacity(declared_expirations.len());
+        let mut expiration_groups = Vec::<SectorExpirationSet>::with_capacity(sectors.len());
 
-        for (&expiration, _) in declared_expirations.iter() {
-            let es = self.may_get(expiration)?;
+        for sector in sectors {
+            // precheck to save unnecessary call to `for_each_while_ranged` below
+            if !all_remaining.contains(&sector.sector_number) {
+                continue;
+            }
+            // scan [sector.expiration, sector.expiration+EPOCHS_IN_DAY] for active sectors.
+            // since a sector may have been moved from another deadline, the possible range for an active sector is [sector.expiration, sector.expiration+EPOCHS_IN_DAY]
+            let end_epoch = (sector.expiration + EPOCHS_IN_DAY) as u64;
+            self.amt.for_each_while_ranged(Some(sector.expiration as u64), None, |epoch, es| {
+                if epoch > end_epoch || !all_remaining.contains(&sector.sector_number) {
+                    // no need to scan for this sector any more
+                    return Ok(false);
+                }
 
-            let group = group_expiration_set(
-                sector_size,
-                &sectors_by_number,
-                &mut all_remaining,
-                es,
-                expiration,
-            );
-            if !group.sector_epoch_set.sectors.is_empty() {
+                if !es.on_time_sectors.get(sector.sector_number) {
+                    // continue scan for this sector
+                    return Ok(true);
```

Comment:
Can stop without even trying to move to the next entry.

```suggestion
                    return Ok(epoch < end_epoch);
```

Reply:
Done.